sirepo/pkcli/test_http.py

```diff
@@ -322,60 +288,230 @@ def _uri(self, uri):
     )
 
 
-async def _run(email, sim_type):
-    await _run_sequential_parallel(
-        await _Client(email=email, sim_type=sim_type).login(),
-    )
+async def _main():
+    async def _get_apps(clients):
+        a = []
+        for c in clients:
+            for t in (
+                    'elegant',
+                    'jspec',
+                    'srw',
+                    'synergia',
+                    'warppba',
+                    'warpvnd',
+            ):
+                a.append(await _App(t, c.copy()).setup_sim_data())
+        return a
+
+    async def _get_clients(*users):
+        l = []
+        for u in users:
+            l.append(_Client(u).login())
+        return await asyncio.gather(*l)
+
+    def _get_sims(apps):
+        s = []
+        for a in apps:
+            t = _CODES[a.sim_type]
+            e = t[random.randrange(len(t))]
+            random.shuffle(e.reports)
+            for r in e.reports:
+                s.append(_Sim(a, e.name, r).run())
+        return s
 
+    def _register_signal_handlers(main_task):
+        def _s(*args):
+            main_task.cancel()
+        signal.signal(signal.SIGTERM, _s)
+        signal.signal(signal.SIGINT, _s)
 
-async def _run_all():
-    l = []
-    for a in (
-            ('one@b.c', 'elegant'),
-            ('one@b.c', 'jspec'),
-            ('one@b.c', 'srw',),
-            ('one@b.c', 'synergia'),
-            ('one@b.c', 'warppba'),
-            ('one@b.c', 'warpvnd'),
-            ('two@b.c', 'elegant'),
-            ('two@b.c', 'jspec'),
-            ('two@b.c', 'srw',),
-            ('two@b.c', 'synergia'),
-            ('two@b.c', 'warppba'),
-            ('two@b.c', 'warpvnd'),
-            ('three@b.c', 'elegant'),
-            ('three@b.c', 'jspec'),
-            ('three@b.c', 'srw',),
-            ('three@b.c', 'synergia'),
-            ('three@b.c', 'warppba'),
-            ('three@b.c', 'warpvnd'),
-    ):
-        l.append(_run(*a))
-    await _cancel_on_exception(asyncio.gather(*l))
-
-
-async def _run_sequential_parallel(client):
-    c = []
-    s = CODES[client.sim_type]
-    e = s[random.randrange(len(s))]
-    random.shuffle(e.reports)
-    for r in e.reports:
-        c.append(client.sim_run(e.name, r))
-    await _cancel_on_exception(asyncio.gather(*c))
-
-
-@contextlib.contextmanager
-def _timer(description):
-    s = time.time()
-    yield
-    if 'run-status' not in description:
-        pkdlog('{} elapsed_time={}', description, time.time() - s)
+    s = _get_sims(
+        await _get_apps(
+            await _get_clients('one@b.c', 'two@b.c', 'three@b.c'),
+        ),
+    )
+    try:
+        t = asyncio.gather(*s)
+        _register_signal_handlers(t)
+        await t
+    except Exception as e:
+        await _cancel_all_tasks(s)
+        if isinstance(e, asyncio.CancelledError):
+            # Will only be cancelled by a signal handler
+            sys.exit(1)
```

Comment:
I don't like sys.exit, but I think I understand why it is here. Is there not a way to simply exit tornado?

Reply:
I used `sys.exit(1)` so we would have a non-zero exit code after a C-c.  If you prefer it could just be `return`. AFAIK no ioloop cleanup has to happen.

Reply:
I don't know that we care about the return code from a sigint or term. I do care about sys.exits in the code as they make the code non-modular. Rather use a return